### DISCUSSION TITLE
Dialogs

### DISCUSSION THREAD

**Discussion by VladislavAntonyuk** [CommunityToolkit/Maui#381] [New Feature Discussions]
It's a new category which I propose to add.
It is more suitable for MAUI Essentials, but features come to Community Toolkit much easier and faster. 

So what this category may include:
- SaveFileDialog
- OpenDirectoryDialog

**Comment by GabrieleMessina** [comment]
That would be great, i've had to implement something similar recently and i think it's a bad miss in MAUI not to have those. 
I think though they should be called FileSavePicker and FolderPicker because that's how Windows SDK calls them and because there is already a FilePicker on Essentials that is conceptually near to these and would be good they have a similar name.

**Comment by pictos** [comment]
Doesn't Essentials have this API? If so we should coordinate with them to avoid duplicate code and/or causing confusion on users (with API should I use?)

**Reply by TheCodeTraveler** [reply]
It does sound similar to [`FilePicker`](https://github.com/dotnet/maui/blob/main/src/Essentials/src/FilePicker/FilePicker.shared.cs), but I don't think `IFilePicker` allows the user to save files, only read files.

@VladislavAntonyuk Could you add additional description about what these APIs would do?

**Reply by VladislavAntonyuk** [reply]
The idea comes from WPF because .NET MAUI becomes a truly cross-platform framework with desktop support.
SaveFileDialog: https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.savefiledialog?view=windowsdesktop-6.0

![image](https://user-images.githubusercontent.com/33021114/165175325-16642a71-8dd6-49db-97dc-55a7cbd97bc0.png)

FolderBrowserDialog: https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.folderbrowserdialog?redirectedfrom=MSDN&view=windowsdesktop-6.0.

They can probably be combined into a single FolderPicker.

Use cases:
1. Select the target directory to save the file.
2. Open the directory to search for files.

Possible API:
```csharp
static Task<Folder> PickAsync(CancellationToken cancellationToken);

class Folder
{
    string Path {get;set;}
    string Name {get;set;}
}
```

**Reply by pictos** [reply]
@VladislavAntonyuk How this is different from this?
https://docs.microsoft.com/en-us/xamarin/essentials/file-picker?tabs=android

Like the API that you described here is to Pick a file, if it was to Save would make more sense, since it's not implemented in Essentials. I can see value in providing features that MAUI.Essentials don't have, that way we can improve that in a faster loop and promote it to .NET MAUI.

That said what do you think about implementing the Save API?

**Reply by VladislavAntonyuk** [reply]
FilePicker works only for Images (even with file options Default).

FolderPicker is for directories.

No, we do not need Save Api, because we can use System.IO. But we need to select the path. So FolderPicker or whatever name we choose is a dialog to select the folder. Its task is to return the path.

**Comment by hevey** [comment]
I like this idea also. I wanted to mention as there was discussion about how folder picking looks on iOS on the community standup stream, here is a screenshot of the Textastic application allowing the user to pick a folder so the app has access to it.

<img width="908" alt="image" src="https://user-images.githubusercontent.com/7008490/167286650-d5d5c5f7-6687-44f1-a39b-ccc4724bd017.png">

**Reply by pictos** [reply]
@hevey do you know if iOS support the action to save a file using this same UI/feature?

**Reply by VladislavAntonyuk** [reply]
@pictos I think it can be achieved with a Share API which is the part of MAUI. Share->Files->Select directory.

**Reply by hevey** [reply]
Good question @pictos. I don't believe so. I would be happy to put a small sample application together to test out the functionality for iOS if there isn't one already.

**Reply by pictos** [reply]
@VladislavAntonyuk but the share API doesn't save the file, right? 

@hevey if you have time to do so that will be great and will help us to design this API better (:

**Reply by VladislavAntonyuk** [reply]
It copies the file from Application folder to the target location

**Comment by rasta-mouse** [comment]
I would really love a save file dialogue.

**Comment by mducharm** [comment]
I wanted to use Blazor Hybrid for a side project, and elected to use the MAUI flavor. In hindsight I might've used WPF instead due to the missing file dialog option. If we could get this added to MAUI, I think it'd be a great addition.

**Reply by bijington** [reply]
@jfversluis have we discussed the idea of a Blazor based toolkit?

**Reply by jfversluis** [reply]
When naming the different toolkits it was definitely considered an option. But I don’t think we have any immediate plans for starting one 😄

**Reply by VladislavAntonyuk** [reply]
As for me Blazor has awesome frameworks like MudBlazor with a lot of controls and alerts.
Dialogs are the integration with system rather then a control, so it can be a part of CommunityToolkit.MAUI or ideally .NET MAUI

**Reply by mducharm** [reply]
The [awesome-blazor repo](https://github.com/AdrienTorris/awesome-blazor) has a list of some Blazor component libraries, for anyone curious.

But yeah, to be more specific about my comment above, I was thinking about [Blazor Hybrid for MAUI](https://docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/tutorials/maui?view=aspnetcore-6.0) vs [Blazor Hybrid for WPF](https://docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/tutorials/wpf?view=aspnetcore-6.0).

**Comment by justcivah** [comment]
I would love a save file dialog too

**Comment by deviprasad987** [comment]
SaveFileDialog is Required

**Comment by marcojak** [comment]
It would be really great to have them. Let's be honest, This is a feature that is required in almost every app (on Windows).
hopefully it will be added soon, otherwise MAUI will never be truly cross platform.

**Comment by VladislavAntonyuk** [comment]
You can track the progress here: https://github.com/CommunityToolkit/Maui/pull/699
